File: algorithm/leetcode5430.design-browser-history.cpp
class BrowserHistory {
    stack<string> bk, ft;
    string cur, ans;
public:
    BrowserHistory(string homepage) {
    	cur = homepage;
    }
    
    void visit(string url) {
    	while(!ft.empty())
    		ft.pop();
    	bk.push(cur);
    	cur = url;
    }
    
    string back(int steps) {
    	ft.push(cur);
    	while(--steps && !bk.empty())
    	{
    		ft.push(bk.top());
    		bk.pop();
    	}
    	if(!bk.empty())
    	{
    		cur = bk.top();
    		bk.pop();
    	}
    	else
    	{
    		cur = ft.top();
    		ft.pop();
    	}
    	return cur;
    }
    
    string forward(int steps) {
    	bk.push(cur);
    	while(--steps && !ft.empty())
    	{
    		bk.push(ft.top());
    		ft.pop();
    	}
    	if(!ft.empty())
    	{
    		cur = ft.top();
    		ft.pop();
    	}
    	else
    	{
    		cur = bk.top();
    		bk.pop();
    	}
    	return cur;
    }
};

File: algorithm/Python3_version/leetcode5430.design-browser-history.py
class BrowserHistory:
    def __init__(self, homepage: str):
        self.pos = 0
        self.stk = [homepage]

    def visit(self, url: str) -> None:
        self.pos += 1
        self.stk = self.stk[0:self.pos]
        self.stk.append(url) 

    def back(self, steps: int) -> str:
        self.pos = max(0, self.pos-steps)
        return self.stk[self.pos]

    def forward(self, steps: int) -> str:
        self.pos = min(len(self.stk)-1, self.pos+steps)
        return self.stk[self.pos]